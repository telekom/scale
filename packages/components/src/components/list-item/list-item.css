/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host(scale-list-item) {
  --display: block;
  --font-size: var(--telekom-typography-font-size-body);
  --line-height: var(--telekom-typography-line-spacing-standard);
  --spacing-left: calc(0.25ch + var(--telekom-spacing-unit-x4));

  --font-size-marker-ordered: var(--telekom-typography-font-size-body);
  --font-weight-marker-ordered: var(--telekom-typography-font-weight-bold);
  --font-size-marker-ordered-nested: var(--telekom-typography-font-size-small);
  --line-height-marker-ordered-nested: var(
    --telekom-typography-line-spacing-standard
  );
  --spacing-top-marker-ordered-nested: calc(
    var(--telekom-spacing-unit-x4) - var(--telekom-spacing-unit-x3)
  );

  --spacing-right-no-marker: var(--telekom-spacing-unit-x2);

  display: var(--display);
}

.list-item {
  position: relative;
  font-size: var(--font-size);
  line-height: var(--line-height);
  padding-left: var(--spacing-left);
}

.list-item:before {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
}

.list-item--unordered:before {
  top: 0.5em;
  border: var(--telekom-spacing-unit-x025) solid currentColor;
  display: block;
  transform: scale(0.66);
  background: currentColor;
  box-sizing: border-box;
  border-radius: 50%;
  width: var(--telekom-spacing-unit-x2);
  height: var(--telekom-spacing-unit-x2);
}

.list-item--nested.list-item--unordered:before {
  border: var(--telekom-spacing-unit-x025) solid currentColor;
  background: transparent;
}

.list-item--ordered:before {
  content: attr(data-index) '.';
  font-size: var(--font-size-marker-ordered);
  font-weight: var(--font-weight-marker-ordered);
}

.list-item--nested.list-item--ordered:before {
  font-size: var(--font-size-marker-ordered-nested);
  line-height: var(--line-height-marker-ordered-nested);
  padding-top: var(--spacing-top-marker-ordered-nested);
}

.list-item--no-marker {
  display: inline-flex;
  align-items: center;
  padding-left: 0;
}

.list-item--no-marker:before {
  display: none;
}

.list-item--no-marker > ::slotted(*) {
  margin-right: var(--spacing-right-no-marker);
}
