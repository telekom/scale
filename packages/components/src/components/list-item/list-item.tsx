/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, h, Prop, Host, Element, State } from '@stencil/core';
import classNames from 'classnames';

/**
 * @todo styles for custom icon (no-marker prop?)
 * @see https://github.com/carbon-design-system/carbon-web-components/tree/master/src/components/list
 */

@Component({
  tag: 'scale-list-item',
  styleUrl: './list-item.css',
  shadow: true,
})
export class ListItem {
  @Element() el: HTMLElement;

  /** Whether this is a child of an ordered scale-list, gets set automatically by its parent */
  @Prop() ordered?: boolean = false;
  /** Index number, useful only for styling the `ordered` type */
  @Prop({ mutable: true }) index?: number;
  /** If `false`, no marker or left padding will be visible */
  @Prop() marker: boolean = true;

  @State() hasNestedChild: boolean = false;
  @State() isNested: boolean = false;

  componentWillLoad() {
    this.isNested = this.isNestedCheck();
  }

  connectedCallback() {
    if (!this.el.hasAttribute('role')) {
      this.el.setAttribute('role', 'listitem');
    }
  }

  handleSlotChange = ({ target }: Event) => {
    this.hasNestedChild =
      (target as HTMLSlotElement).assignedNodes().length > 0;
    this.isNested = this.isNestedCheck();
  };

  isNestedCheck = (): boolean => {
    return this.el.closest('scale-list[slot="nested"]') != null;
  };

  render() {
    return (
      <Host>
        <div
          class={this.getCssClassMap()}
          data-index={this.index}
          part={classNames(
            'base',
            this.ordered ? 'ordered' : 'unordered',
            this.isNested && 'nested',
            !this.marker && 'no-marker'
          )}
        >
          <slot />
          <div
            class="list-item__nested-list"
            part="nested-list"
            hidden={!this.hasNestedChild}
          >
            <slot name="nested" onSlotchange={this.handleSlotChange}></slot>
          </div>
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    const orderType = this.ordered ? 'ordered' : 'unordered';

    return classNames(
      'list-item',
      this.isNested && 'list-item--nested',
      `list-item--${orderType}`,
      !this.marker && 'list-item--no-marker'
    );
  }
}
