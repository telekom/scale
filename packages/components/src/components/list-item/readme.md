# scale-list-item



<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description                                                                            | Type      | Default     |
| --------- | --------- | -------------------------------------------------------------------------------------- | --------- | ----------- |
| `index`   | `index`   | Index number, useful only for styling the `ordered` type                               | `number`  | `undefined` |
| `marker`  | `marker`  | If `false`, no marker or left padding will be visible                                  | `boolean` | `true`      |
| `ordered` | `ordered` | Whether this is a child of an ordered scale-list, gets set automatically by its parent | `boolean` | `false`     |


## Shadow Parts

| Part            | Description |
| --------------- | ----------- |
| `"nested-list"` |             |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
