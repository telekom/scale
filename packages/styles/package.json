{
    "name": "@telements/styles",
    "version": "0.0.1",
    "main": "dist/index.js",
    "scripts": {
        "build": "tsc -b",
        "postbuild": "node ./dist/export-css.js"
    },
    "author": "@eeegor <egor.kirpichev@telekom.de>",
    "license": "MIT",
    "devDependencies": {
        "@types/jest": "^24.0.13",
        "jest": "^24.8.0",
        "typescript": "^3.4.5"
    },
    "dependencies": {
        "@types/classnames": "^2.2.8",
        "@types/tinycolor2": "^1.4.2",
        "classnames": "^2.2.6",
        "tinycolor2": "^1.4.1"
    }
}
