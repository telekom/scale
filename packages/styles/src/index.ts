import { button, buttonClassNames } from './button';
import { alert } from './alert';
import { card } from './card';

export {
	alert,
	button,
	buttonClassNames,
	card,
}
