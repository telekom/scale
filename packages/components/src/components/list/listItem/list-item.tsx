import { Component, h, Prop, Host } from '@stencil/core';
import { CssClassMap } from '../../../utils/utils';
import classNames from 'classnames';
import { styles } from './list-item.styles';
import { CssInJs } from '../../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../../utils/base-interface';
@Component({
  tag: 'scale-list-item',
  shadow: true,
})
export class ListItem implements Base {
  /** (optional) List item icon */
  @Prop() icon?: string;
  /** (optional) list Icon size */
  @Prop() iconSize?: number = 16;
  /** (optional) List item style type */
  @Prop() type?: string;
  /** (optional) Injected jss styles */
  @Prop() styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('ListItem', styles) stylesheet: StyleSheet;
  componentWillLoad() {}
  componentDidUnload() {}
  componentWillUpdate() {}

  render() {
    this.stylesheet.update({
      type: this.type,
    });

    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <li class={this.getCssClassMap()}>
          {!!this.icon && (
            <scale-icon path={this.icon} size={this.iconSize}></scale-icon>
          )}
          <slot />
        </li>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes['list-item'],
      this.type && classes['list-item--type'],
      !!this.icon && classes[`list-item--has-icon`]
    );
  }
}
