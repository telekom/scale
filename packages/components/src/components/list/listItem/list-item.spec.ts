import { newSpecPage } from '@stencil/core/testing';
import { ListItem } from './list-item';
import { styles } from './list-item.styles';
import jss from 'jss';

describe('ListItem', () => {
  let element;
  let stylesheet;
  beforeEach(async () => {
    element = new ListItem();
    stylesheet = element.stylesheet = jss.createStyleSheet(styles as any);
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [ListItem],
      html: `<scale-list-item>default</scale-list-item>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should handle css classes', () => {
    element.type = 'circle';
    stylesheet.addRule('list-item--type', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['list-item--type']
    );
  });
});
