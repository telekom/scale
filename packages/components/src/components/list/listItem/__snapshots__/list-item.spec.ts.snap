// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should match snapshot 1`] = `
<scale-list-item>
  <mock:shadow-root>
    <style>
      .list-item-0-0-4 {}
.list-item--type-0-0-5 {
  list-style: square;
}
.list-item--has-icon-0-0-6 {
  display: flex;
  list-style: none !important;
  align-items: center;
  margin-left: -1.5rem;
}
    </style>
    <li class="list-item-0-0-4">
      <slot></slot>
    </li>
  </mock:shadow-root>
  default
</scale-list-item>
`;
